File: CartQuestiOS/CartQuestiOS/Views/Cart/CartBuilderViewModel.swift
import Foundation
import Observation

@Observable
class CartBuilderViewModel {
    var searchQuery: String = "" {
        didSet { debouncedSearch() }
    }
    var searchResults: [KrogerProduct] = []
    var isSearching: Bool = false
    var cart: Cart = Cart()
    var isSaving: Bool = false
    var addingSubstituteForIndex: Int?

    private let krogerService: KrogerService
    private let cartRepository: CartRepository
    private var searchTask: Task<Void, Never>?
    private var saveTask: Task<Void, Never>?

    init(
        krogerService: KrogerService = KrogerService(
            clientId: "KROGER_CLIENT_ID_PLACEHOLDER",
            clientSecret: "KROGER_CLIENT_SECRET_PLACEHOLDER"
        ),
        cartRepository: CartRepository = CartRepository()
    ) {
        self.krogerService = krogerService
        self.cartRepository = cartRepository
        Task { await loadActiveCart() }
    }

    // MARK: - Load

    private func loadActiveCart() async {
        do {
            if let activeCart = try await cartRepository.getActiveCart() {
                cart = activeCart
            }
        } catch {
            print("Failed to load active cart: \(error)")
        }
    }

    // MARK: - Search

    private func debouncedSearch() {
        searchTask?.cancel()
        let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else {
            searchResults = []
            isSearching = false
            return
        }
        searchTask = Task {
            try? await Task.sleep(for: .milliseconds(400))
            guard !Task.isCancelled else { return }
            await search(query: query)
        }
    }

    func search() {
        let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else { return }
        searchTask?.cancel()
        searchTask = Task { await search(query: query) }
    }

    private func search(query: String) async {
        isSearching = true
        defer { isSearching = false }
        do {
            let results = try await krogerService.searchProducts(term: query)
            guard !Task.isCancelled else { return }
            searchResults = results
        } catch {
            guard !Task.isCancelled else { return }
            print("Search failed: \(error)")
            searchResults = []
        }
    }

    // MARK: - Cart mutations

    func addToCart(product: KrogerProduct) {
        if let subIndex = addingSubstituteForIndex {
            addSubstitute(cartItemIndex: subIndex, product: product)
            addingSubstituteForIndex = nil
        } else {
            let imageUrl = product.images.first?.sizes.first?.url ?? ""
            let item = CartItem(
                productId: product.productId,
                name: product.description,
                brand: product.brand,
                imageUrl: imageUrl,
                quantity: 1,
                substitutes: []
            )
            cart.items.append(item)
        }
        searchQuery = ""
        searchResults = []
        debounceSave()
    }

    func removeFromCart(at index: Int) {
        guard cart.items.indices.contains(index) else { return }
        cart.items.remove(at: index)
        debounceSave()
    }

    func addSubstitute(cartItemIndex: Int, product: KrogerProduct) {
        guard cart.items.indices.contains(cartItemIndex) else { return }
        let sub = Substitute(
            productId: product.productId,
            name: product.description,
            brand: product.brand
        )
        cart.items[cartItemIndex].substitutes.append(sub)
        debounceSave()
    }

    func removeSubstitute(cartItemIndex: Int, subIndex: Int) {
        guard cart.items.indices.contains(cartItemIndex),
              cart.items[cartItemIndex].substitutes.indices.contains(subIndex) else { return }
        cart.items[cartItemIndex].substitutes.remove(at: subIndex)
        debounceSave()
    }

    func updateQuantity(at index: Int, quantity: Int) {
        guard cart.items.indices.contains(index) else { return }
        cart.items[index].quantity = max(1, quantity)
        debounceSave()
    }

    // MARK: - Persistence

    private func debounceSave() {
        saveTask?.cancel()
        saveTask = Task {
            try? await Task.sleep(for: .milliseconds(800))
            guard !Task.isCancelled else { return }
            await saveCart()
        }
    }

    private func saveCart() async {
        isSaving = true
        defer { isSaving = false }
        do {
            cart = try await cartRepository.saveCart(cart)
        } catch {
            print("Failed to save cart: \(error)")
        }
    }
}

File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/ui/cart/CartBuilderViewModel.kt
package com.amoghghadge.cartquestandroid.ui.cart

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.amoghghadge.cartquestandroid.BuildConfig
import com.amoghghadge.cartquestandroid.data.model.Cart
import com.amoghghadge.cartquestandroid.data.model.CartItem
import com.amoghghadge.cartquestandroid.data.model.KrogerProduct
import com.amoghghadge.cartquestandroid.data.model.Substitute
import com.amoghghadge.cartquestandroid.data.remote.KrogerApiService
import com.amoghghadge.cartquestandroid.data.remote.KrogerAuthManager
import com.amoghghadge.cartquestandroid.data.repository.CartRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class CartBuilderUiState(
    val searchQuery: String = "",
    val searchResults: List<KrogerProduct> = emptyList(),
    val isSearching: Boolean = false,
    val cart: Cart = Cart(),
    val isSaving: Boolean = false,
    val addingSubstituteForIndex: Int? = null
)

class CartBuilderViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(CartBuilderUiState())
    val uiState: StateFlow<CartBuilderUiState> = _uiState.asStateFlow()

    private val cartRepository = CartRepository()
    private val krogerApiService = KrogerApiService.create()
    private val krogerAuthManager = KrogerAuthManager(
        clientId = BuildConfig.KROGER_CLIENT_ID,
        clientSecret = BuildConfig.KROGER_CLIENT_SECRET
    )

    private var searchJob: Job? = null
    private var saveJob: Job? = null

    init {
        viewModelScope.launch {
            try {
                val activeCart = cartRepository.getActiveCart()
                if (activeCart != null) {
                    _uiState.update { it.copy(cart = activeCart) }
                }
            } catch (_: Exception) {
                // No active cart found, keep default empty cart
            }
        }
    }

    fun search(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
        searchJob?.cancel()
        if (query.isBlank()) {
            _uiState.update { it.copy(searchResults = emptyList(), isSearching = false) }
            return
        }
        searchJob = viewModelScope.launch {
            delay(500)
            _uiState.update { it.copy(isSearching = true) }
            try {
                val token = krogerAuthManager.getToken()
                val response = krogerApiService.searchProducts(
                    auth = "Bearer $token",
                    term = query
                )
                _uiState.update { it.copy(searchResults = response.data, isSearching = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isSearching = false) }
            }
        }
    }

    fun addToCart(product: KrogerProduct) {
        val substituteIndex = _uiState.value.addingSubstituteForIndex
        if (substituteIndex != null) {
            addSubstitute(substituteIndex, product)
            cancelAddingSubstitute()
            return
        }
        val imageUrl = product.images.firstOrNull()?.sizes?.firstOrNull()?.url.orEmpty()
        val newItem = CartItem(
            productId = product.productId,
            name = product.description,
            brand = product.brand,
            imageUrl = imageUrl,
            quantity = 1
        )
        _uiState.update { state ->
            state.copy(
                cart = state.cart.copy(items = state.cart.items + newItem),
                searchQuery = "",
                searchResults = emptyList()
            )
        }
        debounceSave()
    }

    fun removeFromCart(index: Int) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList().apply { removeAt(index) }
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun addSubstitute(cartItemIndex: Int, product: KrogerProduct) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            val item = items[cartItemIndex]
            val substitute = Substitute(
                productId = product.productId,
                name = product.description,
                brand = product.brand
            )
            items[cartItemIndex] = item.copy(substitutes = item.substitutes + substitute)
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun removeSubstitute(cartItemIndex: Int, subIndex: Int) {
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            val item = items[cartItemIndex]
            val subs = item.substitutes.toMutableList().apply { removeAt(subIndex) }
            items[cartItemIndex] = item.copy(substitutes = subs)
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun updateQuantity(index: Int, quantity: Int) {
        if (quantity < 1) return
        _uiState.update { state ->
            val items = state.cart.items.toMutableList()
            items[index] = items[index].copy(quantity = quantity)
            state.copy(cart = state.cart.copy(items = items))
        }
        debounceSave()
    }

    fun startAddingSubstitute(index: Int) {
        _uiState.update { it.copy(addingSubstituteForIndex = index) }
    }

    fun cancelAddingSubstitute() {
        _uiState.update { it.copy(addingSubstituteForIndex = null) }
    }

    private fun debounceSave() {
        saveJob?.cancel()
        saveJob = viewModelScope.launch {
            delay(1000)
            _uiState.update { it.copy(isSaving = true) }
            try {
                val savedCart = cartRepository.saveCart(_uiState.value.cart)
                _uiState.update { it.copy(cart = savedCart, isSaving = false) }
            } catch (_: Exception) {
                _uiState.update { it.copy(isSaving = false) }
            }
        }
    }
}
